### PR TITLE
(Feature) #1020 User should receive hanuka miracle bonus

### PR DESCRIPTION
**Pull request by yaroslav-fedyshyn-nordwhale**
# Description

Add endpoint which verifying whether the user gets a bonus on a specific day or not. If not then send the request to the blockchain to award the user with Hanuka bonus.

The link to the GoodDapp PR - https://github.com/GoodDollar/GoodDAPP/pull/1059

About https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/gooddapp/1020

# How Has This Been Tested?

Described in GoodDaap PR.

Need to set desired env var

Optional. Default: 23
HANUKA_START_DAY

Optional. Default: 30
HANUKA_END_DAY

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: src/server/verification/verificationAPI.js [hunk 1]
```diff
@@ -390,6 +391,94 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
+    '/verify/hanuka-bonus',
+    passport.authenticate('jwt', { session: false }),
+    wrapAsync(async (req, res, next) => {
+      const log = req.log.child({ from: 'verificationAPI - verify/hanuka-bonus' })
+      const { user } = req
+      const now = moment().utcOffset('+0200')
+      const startHanuka = moment(`${conf.hanukaStartDay}/12`, 'DD/MM')
+      const endHanuka = moment(`${conf.hanukaEndDay}/12`, 'DD/MM').endOf('day')
+
+      log.info('Current date', now)
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: put more thought into what you log as info/debug, merge logs

File: src/server/verification/verificationAPI.js [hunk 2]
```diff
@@ -390,6 +391,94 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
+
+      if (startHanuka.isAfter(now) || now.isAfter(endHanuka)) {
+        log.info('That is no the period of Hanuka bonus')
+
+        return res.json({
+          ok: 1
+        })
+      }
+
+      const startsFromDay = 23
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: we have env for that

File: src/server/verification/verificationAPI.js [hunk 3]
```diff
@@ -390,6 +391,94 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
+      const currentDay = now.format('DD')
+      const currentDayNumber = Number(currentDay - startsFromDay + 1)
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: use days diff of moment, your calculation can be wrong

File: src/server/verification/verificationAPI.js [hunk 4]
```diff
@@ -390,6 +391,94 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
+      const bonusInWei = gdToWei(currentDayNumber)
+
+      log.info('Bonus in wei', {
+        bonusInWei,
+        bonus: currentDayNumber
+      })
+
+      const { release, fail } = await txManager.lock(user.gdAddress, 0)
+
+      AdminWallet.redeemBonuses(user.gdAddress, bonusInWei, {
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: return result to user on txhash, so we dont keep connection open for long time.
no need to return result onerror and onreceipt

File: src/server/verification/verificationAPI.js [hunk 5]
```diff
@@ -390,6 +391,94 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
+   * @ignore
+   */
+  app.get(
+    '/verify/hanuka-bonus',
+    passport.authenticate('jwt', { session: false }),
+    wrapAsync(async (req, res, next) => {
+      const log = req.log.child({ from: 'verificationAPI - verify/hanuka-bonus' })
+      const { user } = req
+      const now = moment().utcOffset('+0200')
+      const startHanuka = moment(`${conf.hanukaStartDay}/12`, 'DD/MM')
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: lets use full date in env mmddyy

File: src/server/verification/verificationAPI.js (hunk 1 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: Done

File: src/server/verification/verificationAPI.js (hunk 2 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: done

File: src/server/verification/verificationAPI.js (hunk 3 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: done

File: src/server/verification/verificationAPI.js (hunk 4 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: done

File: src/server/verification/verificationAPI.js (hunk 5 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: done

File: src/server/verification/verificationAPI.js [hunk 6]
```diff
@@ -444,6 +444,11 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
       const { release, fail } = await txManager.lock(user.gdAddress, 0)
 
       AdminWallet.redeemBonuses(user.gdAddress, bonusInWei, {
+        onTransactionHash: r => {
+          return res.status(200).json({
+            ok: 1
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: return txhash also

File: src/server/verification/verificationAPI.js [hunk 7]
```diff
@@ -456,20 +461,11 @@ const setup = (app: Router, verifier: VerificationAPI, storage: StorageAPI) => {
...
           release()
-
-          return res.status(200).json({
-            ok: 1
-          })
         },
         onError: e => {
           log.error('Bonuses charge failed', e.message, e, user)
 
           fail()
```
Comment (sirpy) [inline src/server/verification/verificationAPI.js]: maybe check if connection is still open and send json result if it is, maybe ontxhash wont be called

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale @AlexeyKosinski i merged it accidentally, left a review with some changes required

File: src/server/verification/verificationAPI.js (hunk 6 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/verificationAPI.js]: Done

**Comment by yaroslav-fedyshyn-nordwhale** [comment]
> @yaroslav-fedyshyn-nordwhale @AlexeyKosinski i merged it accidentally, left a review with some changes required

Made changes and add them to the new PR. Also tests added. Link to the new PR - https://github.com/GoodDollar/GoodServer/pull/135

**Pull request merged**